### ISSUE TITLE
Trying to connect to SQL Server 2014 results in '[BUG] Segmentation fault'

### ISSUE THREAD

**Issue by thuringian** [rails-sqlserver/tiny_tds#246]
I've been using this gem for a while successfully with SQL 2008 R2 servers. I'm now testing trying to connect to a SQL 2014 server on the same network and it fails, throwing a Segmentation fault error. If I change my connection info for one of our SQL 2008 R2 servers the same code works fine. Am I missing something for using this with SQL 2014?

`require 'tiny_tds'`
`conn = TinyTds::Client.new(:username => 'user', :password => 'password', :host => 'host')`

error output:

```
/var/lib/gems/2.1.0/gems/tiny_tds-0.7.0/lib/tiny_tds/client.rb:74: [BUG] Segmentation fault at 0x007ffeb7d43fe8
ruby 2.1.2p95 (2014-05-08) [x86_64-linux-gnu]

-- Control frame information -----------------------------------------------
c:0005 p:---- s:0017 e:000016 CFUNC  :connect
c:0004 p:0320 s:0013 e:000012 METHOD /var/lib/gems/2.1.0/gems/tiny_tds-0.7.0/lib/tiny_tds/client.rb:74 [FINISH]
c:0003 p:---- s:0009 e:000008 CFUNC  :new
c:0002 p:0059 s:0005 E:0021d0 EVAL   test.rb:9 [FINISH]
c:0001 p:0000 s:0002 E:0009f8 TOP    [FINISH]

-- Ruby level backtrace information ----------------------------------------
test.rb:9:in `<main>'
test.rb:9:in `new'
/var/lib/gems/2.1.0/gems/tiny_tds-0.7.0/lib/tiny_tds/client.rb:74:in `initialize'
/var/lib/gems/2.1.0/gems/tiny_tds-0.7.0/lib/tiny_tds/client.rb:74:in `connect'

-- C level backtrace information -------------------------------------------
/usr/lib/x86_64-linux-gnu/libruby-2.1.so.2.1(+0x18a4b7) [0x7f7b8c4f14b7]
/usr/lib/x86_64-linux-gnu/libruby-2.1.so.2.1(+0x18a583) [0x7f7b8c4f1583]
/usr/lib/x86_64-linux-gnu/libruby-2.1.so.2.1(+0x6db13) [0x7f7b8c3d4b13] printf-parse.h:88
/usr/lib/x86_64-linux-gnu/libruby-2.1.so.2.1(rb_bug+0xb3) [0x7f7b8c3d5183] printf-parse.h:75
/usr/lib/x86_64-linux-gnu/libruby-2.1.so.2.1(+0x12090e) [0x7f7b8c48790e] getnetgrent_r.c:486
/lib/x86_64-linux-gnu/libc.so.6(+0x352f0) [0x7f7b8bfd22f0] ../sysdeps/posix/killpg.c:37
```

**Comment by metaskills** [comment]
Could you try bundling to the latest in master and let me know how that works?

**Comment by thuringian** [comment]
Sorry, not sure how to go about trying that. I'm not currently using a gemfile or bundler. Just executing with a single 'test.rb' script to see if I can connect.

Thanks

**Comment by metaskills** [comment]
Damn... what version of FreeTDS are you building against?

**Comment by thuringian** [comment]
using Ubuntu 15.04 and freetds-dev package version 0.91-6build1

**Comment by thuringian** [comment]
I've also confirmed after setting up a named server in freetds.conf I'm able to connect to the server directly with freetds using:

tsql -S server -U user

**Comment by metaskills** [comment]
Hmmm, could you search previous issues with Ubuntu keyword to see if there is anything in there that helps? https://github.com/rails-sqlserver/tiny_tds/search?q=Ubuntu&type=Issues

In short, if you can connect with `tsql` and you remove and reinstall TinyTDS so it compiles against that same version, you should be GOLDEN

**Comment by thuringian** [comment]
Sorry no luck. The closest thing I can find is this issue which describes a similar error but different cause (trying to connect to Azure). I'm not trying to use Azure in this case.

http://stackoverflow.com/questions/11632743/connecting-to-sql-azure-in-rails-using-tiny-tds-bug-segmentation-fault.

I did a gem uninstall of tiny_tds (all versions), and then reinstalled 7.0.

I've also tried upping my ruby version to 2.1.5. I now get a slightly different different error but still seems to be the same issue. Tried both the reinstalled 7.0 version and the bundled version from git.

`~/Dropbox/test/shoppers$ bundle
Updating https://github.com/rails-sqlserver/tiny_tds.git
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Using mini_portile 0.6.2
Using tiny_tds 0.9.0.beta1 (was 0.7.0) from https://github.com/rails-sqlserver/tiny_tds.git (at master)
Using bundler 1.10.6
Bundle complete! 1 Gemfile dependency, 3 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
~/Dropbox/test/shoppers$ ruby test.rb
/home/mike/.bundler/ruby/2.1.0/tiny_tds-ec578988d633/lib/tiny_tds/client.rb:43: stack level too deep (SystemStackError)`

**Comment by metaskills** [comment]
Really sorry this is happening. FWIW, I do all my local testing on 2014. The fact that it works on 2008 is super odd. I think that means we have to start looking at what is different and how the server may be affecting this. For example, when connecting to 2014, can you test connecting to a simple database and new user. Just to remove any potential ways it may be trying to configure the client. To that end, have you tried using setting the `TDSVER=7.3` environment variable before connecting too? Maybe `TDSVER=7.1`? What does `$ tsql -C` show?

**Comment by thuringian** [comment]
Okay...this is really weird. I created a new sql login username `tdsuser` and password `password` and it connects. My only guess is it has something to do with the password string for my other login I was testing with, which is somewhat long, so maybe something in how tiny_tds handles that. I was trying to use `9ad93263f3c38abcf00fd4ce3048c69f` as a password if you want to try to reproduce the issue against a 2014 database. Strange still that freetds was able to connect with it but not tiny_tds.

And to answer your question, yes I had previously tried specifying `:tds_version => "7.0"`, `:tds_version => "7.1"`, `:tds_version => "7.2"`, or `:tds_version => "7.3"` in the connection options.

Just in case here is the output of `tsql -C`

`Compile-time settings (established with the "configure" script)
                            Version: freetds v0.91
             freetds.conf directory: /etc/freetds
     MS db-lib source compatibility: no
        Sybase binary compatibility: yes
                      Thread safety: yes
                      iconv library: yes
                        TDS version: 4.2
                              iODBC: no
                           unixodbc: yes
              SSPI "trusted" logins: no
                           Kerberos: yes`

Really appreciate the help and effort!

**Comment by metaskills** [comment]
That password is 33 chars. We used to have [this patch](https://github.com/rails-sqlserver/tiny_tds/commit/5006f4b397d5b62e78dc5f7658670c85c1ebdaab) in place but I removed it because I thought it as no longer needed. Let me test a few things and get back to you. Really appreciate you hanging in there.

**Comment by metaskills** [comment]
OK... so that 30 char patch was only applied when we built the Windows static libraries. I can see in latest 0.91.112 (NOTE THAT .112 TINY) up to 0.95.latest to 0.99.latest all `#define TDS_MAX_LOGIN_STR_SZ 128`. I switched to my local 0.91.112 FreeTDS install and changed my password in 2014 management studio:

```
$ TINYTDS_UNIT_PASS=9ad93263f3c38abcf00fd4ce3048c69f rake test
```

These ran and worked great. Maybe your 0.91 is not the latest .112 version?
